From b4e4223a7764632756e55345227cc1c6715be0b4 Mon Sep 17 00:00:00 2001
From: Slava Aseev <ptrnine@altlinux.org>
Date: Wed, 11 May 2022 18:40:21 +0300
Subject: [PATCH 2/3] Do not create EntryLocation with empty key

Empty keys are not allowed, so the EntryLocation::fromUniqueLabel()
function now identifies the KWallet folder by the first slash. Also a
label with a single trailing slash will be treated as a key without a
folder.
---
 src/runtime/kwalletd/autotests/fdo_secrets_test.cpp   | 5 +++++
 src/runtime/kwalletd/kwalletfreedesktopattributes.cpp | 2 +-
 src/runtime/kwalletd/kwalletfreedesktopcollection.cpp | 4 ++--
 src/runtime/kwalletd/kwalletfreedesktopservice.cpp    | 4 ++--
 4 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/src/runtime/kwalletd/autotests/fdo_secrets_test.cpp b/src/runtime/kwalletd/autotests/fdo_secrets_test.cpp
index 1c324be..bc8b424 100644
--- a/src/runtime/kwalletd/autotests/fdo_secrets_test.cpp
+++ b/src/runtime/kwalletd/autotests/fdo_secrets_test.cpp
@@ -48,6 +48,11 @@ void FdoSecretsTest::collectionStaticFunctions()
         {{"Passwords", "password__"}, {"Passwords/password__", -1}},
         {{"Passwords__3_", "password__200_"}, {"Passwords__3_/password", 200}},
         {{"", "password"}, {"/password", -1}},
+        {{"", "/"}, {"//", -1}},
+        {{FDO_SECRETS_DEFAULT_DIR, "/"}, {"/", -1}},
+        {{FDO_SECRETS_DEFAULT_DIR, "/__2_"}, {"/", 2}},
+        {{"https:", "/foobar.org/"}, {"https://foobar.org/", -1}},
+        {{"https:", "/foobar.org/__80_"}, {"https://foobar.org/", 80}},
     };
 
     runTestset(
diff --git a/src/runtime/kwalletd/kwalletfreedesktopattributes.cpp b/src/runtime/kwalletd/kwalletfreedesktopattributes.cpp
index b0791cb..e1cba15 100644
--- a/src/runtime/kwalletd/kwalletfreedesktopattributes.cpp
+++ b/src/runtime/kwalletd/kwalletfreedesktopattributes.cpp
@@ -88,7 +88,7 @@ static QString entryLocationToStr(const EntryLocation &entryLocation)
 
 static EntryLocation splitToEntryLocation(const QString &entryLocation)
 {
-    const int slashPos = entryLocation.lastIndexOf(QChar::fromLatin1('/'));
+    const int slashPos = entryLocation.indexOf(QChar::fromLatin1('/'));
     if (slashPos == -1) {
         qCWarning(KWALLETD_LOG) << "Entry location '" << entryLocation << "' has no slash '/'";
         return {};
diff --git a/src/runtime/kwalletd/kwalletfreedesktopcollection.cpp b/src/runtime/kwalletd/kwalletfreedesktopcollection.cpp
index 3d5a5c2..74bc1c5 100644
--- a/src/runtime/kwalletd/kwalletfreedesktopcollection.cpp
+++ b/src/runtime/kwalletd/kwalletfreedesktopcollection.cpp
@@ -262,8 +262,8 @@ EntryLocation KWalletFreedesktopCollection::makeUniqueEntryLocation(const QStrin
 {
     QString dir, name;
 
-    const int slashPos = label.lastIndexOf(QChar::fromLatin1('/'));
-    if (slashPos == -1) {
+    const int slashPos = label.indexOf(QChar::fromLatin1('/'));
+    if (slashPos == -1 || slashPos == label.size() - 1) {
         dir = QStringLiteral(FDO_SECRETS_DEFAULT_DIR);
         name = label;
     } else {
diff --git a/src/runtime/kwalletd/kwalletfreedesktopservice.cpp b/src/runtime/kwalletd/kwalletfreedesktopservice.cpp
index c6b4821..17c66ea 100644
--- a/src/runtime/kwalletd/kwalletfreedesktopservice.cpp
+++ b/src/runtime/kwalletd/kwalletfreedesktopservice.cpp
@@ -73,8 +73,8 @@ EntryLocation EntryLocation::fromUniqueLabel(const FdoUniqueLabel &uniqLabel)
     QString dir;
     QString name = uniqLabel.label;
 
-    const int slashPos = uniqLabel.label.lastIndexOf(QChar::fromLatin1('/'));
-    if (slashPos == -1) {
+    const int slashPos = uniqLabel.label.indexOf(QChar::fromLatin1('/'));
+    if (slashPos == -1 || slashPos == uniqLabel.label.size() - 1) {
         dir = QStringLiteral(FDO_SECRETS_DEFAULT_DIR);
     } else {
         dir = uniqLabel.label.left(slashPos);
-- 
2.36.0

